Persistent Node Ids

bleskes opened the pull request:
Node IDs are currently randomly generated during node startup. That means they change every time the node is restarted. While this doesn't matter for ES proper, it makes it hard for external services to track nodes. Another, more minor, side effect is that indexing the output of, say, the node stats API results in creating new fields due to node ID being used as keys.

The first approach I considered was to use the node's published address as the base for the id. We already [treat nodes with the same address as the same](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java#L387) so this is a simple change (see [here](https://github.com/elastic/elasticsearch/compare/master...bleskes:node_persistent_id_based_on_address)). While this is simple and it works for probably most cases, it is not perfect. For example, if after a node restart, the node is not able to bind to the same port (because it's not yet freed by the OS), it will cause the node to still change identity. Also in environments where the host IP can change due to a host restart, identity will not be the same. 

Due to those limitation, I opted to go with a different approach where the node id will be persisted in the node's data folder. This has the upside of connecting the id to the nodes data. It also means that the host can be adapted in any way (replace network cards, attach storage to a new VM). I

t does however also have downsides - we now run the risk of two nodes having the same id, if someone copies clones a data folder from one node to another. To mitigate this I changed the semantics of the protection against multiple nodes with the same address to be stricter - it will now reject the incoming join if a node exists with the same id but a different address. Note that if the existing node doesn't respond to pings (i.e., it's not alive) it will be removed and the new node will be accepted when it tries another join.

Last, and most importantly, this change requires that *all* nodes persist data to disk. This is a change from current behavior where only data & master nodes store local files. This is the main reason for marking this PR as breaking.

Other less important notes:
- DummyTransportAddress is removed as we need a unique network address per node. Use `LocalTransportAddress.buildUnique()` instead.
- I renamed `node.add_lid_to_custom_path` to `node.add_lock_id_to_custom_path` to avoid confusion with the node ID which is now part of the `NodeEnvironment` logic.
- I removed the `version` paramater from `MetaDataStateFormat#write` , it wasn't really used and was just in the way :)
- TribeNodes are special in the sense that they do start multiple sub-nodes (previously known as client nodes). Those sub-nodes do not store local files and will this have a new id every restart. I have some ideas on how to change this but it will require more sublte changes that I rather be tackled in a follow up issue if needed (/cc @pickypg  @uboness @clintongormley )

ywelsch commented on core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java:
```diff
@@ -376,20 +378,26 @@ public ClusterState execute(ClusterState currentState) {
+                        for (MembershipAction.JoinCallback callback: entry.getValue()) {
+                            joinCallbacksToFail.add(new Tuple<>(callback, failure));
+                        }
+                    } else if (currentState.nodes().nodeExists(node.getId())) {
```
If a node with same persisted node id but different address tries to join the cluster (because the data folder was copied from another node) it will not manage to join the cluster. It will not have any meaningful error message either except that it couldn't join. Can we tell the offending node why it failed using joinCallbacksToFail?

bleskes replied:
That's part of the exception the joining nodes gets? It will be logged [here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java#L488)

bleskes replied:
Oh - I see what you're saying, this should be caught by the pre-flight check on the DiscoNodes (which is what I actually wanted to achieve - "it will now reject the incoming join if a node exists with the same id but a different address." - great catch)

ywelsch replied:
One more question: What about `DiscoveryNodes.delta()`? It does not detect if node has been removed and readded. It also does not take `equalsIncludingMetaData` into account. This means that other code that relies upon the nodesDelta API in ClusterChangedEvent to detect if node has restarted might miss a restart? One such example is possibly `SnapshotsService.removedNodesCleanupNeeded`.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java:
```diff
@@ -565,6 +566,10 @@ public Builder(DiscoveryNodes nodes) {
         public Builder put(DiscoveryNode node) {
+            final String preflight = preflightPut(node);
+            if (preflight != null) {
+                throw new ElasticsearchException(preflight);
```
IllegalArgumentException?

bleskes replied:
make sense.

ywelsch commented on core/src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -128,23 +133,31 @@ public String toString() {
+
+
+    public static final Setting<Long> NODE_ID_SEED_SETTING =
+        // don't use node.id.seed so it won't be seen as an attribute
```
I vaguely remember that attributes have been moved to attrs. namespace? Can this finally become `node.id.seed`?

bleskes replied:
yes it can!

ywelsch commented on core/src/main/java/org/elasticsearch/env/NodeMetaData.java:
```diff
@@ -0,0 +1,115 @@
+
+    public NodeMetaData(final String nodeID) {
+        Objects.requireNonNull(nodeID);
+        this.nodeID = nodeID;
```
this is usually written
```
this.nodeId = Objects.requireNonNull(nodeID);
```
Quick and easy.

ywelsch commented on core/src/main/java/org/elasticsearch/env/NodeMetaData.java:
```diff
@@ -0,0 +1,115 @@
+
+    private static final String NODE_ID_KEY = "node_id";
+
+    private final String nodeID;
```
s/nodeID/nodeId/ everywhere :-)

bleskes replied:
sure

ywelsch commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -138,6 +138,16 @@
+
+    /**
+     * controls whether the node is allowed to persist things like metadata to disk
+     * Note that this does not control whether the node store actual indices (see
```
s/store/stores/

bleskes replied:
done

ywelsch commented on core/src/test/java/org/elasticsearch/discovery/zen/NodeJoinControllerTests.java:
```diff
@@ -413,6 +413,15 @@ public void testNewClusterStateOnExistingNodeJoin() throws InterruptedException,
+        ExecutionException e = expectThrows(ExecutionException.class, () -> joinNode(other_node));
+        assertThat(e.getMessage(), containsString("found existing node"));
+    }
+
```
also add test for situation I described above (same node id but different address)

bleskes replied:
will do

ywelsch commented:
Left a few comments but in general looks good. I hope we won't have to re-add ephemeral node ids at a later point... ;-)

bleskes commented:
@ywelsch thanks. I addressed your comments and pushed some new commits. Sadly, my Jenkins server signaled an issue with our testing cluster. In the name of reproducibility, it captures the initial seeds of each nodes and when it resets the cluster and start up nodes that were shut down but tests, it will re-use that seed so a node will have the same id. However, because how things work elsewhere, this id may be taken by another node that re-used the data folder of the old node. I added an assertion to LocalDiscovery that surfaces this clearly (had to chase it down, which was sneaky). I'm still thinking about a proper solution. Will ping once it's there.

bleskes commented:
@ywelsch I found a simple way to work around the test cluster issues. This is ready for review again...

ywelsch commented on core/src/main/java/org/elasticsearch/env/NodeMetaData.java:
```diff
@@ -56,39 +55,39 @@ public boolean equals(Object o) {
 
-        public void setNodeID(String nodeID) {
-            this.nodeID = nodeID;
+        public void setnodeId(String nodeId) {
```
s/setnodeId/setNodeId/

ywelsch commented on core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java:
```diff
@@ -372,24 +379,31 @@ public ClusterState execute(ClusterState currentState) {
+                        }
+                    } else if (currentState.nodes().nodeExists(node.getId())) {
                         logger.debug("received a join request for an existing node [{}]", node);
+                        joinCallbacksToRespondTo.addAll(entry.getValue());
```
should we update nodesBuilder in this case? What if node attributes /  roles have been changed in the meantime?

bleskes replied:
hmm.. you're right. A quick restart + change of attributes can make this happen. I'll add this and add a test.

ywelsch commented:
superseded by #17987 (which keeps process semantics for DiscoveryNode)
